@planetarium/account-web3-secret-storage/src/Web3KeyStore.ts

```diff
@@ -0,0 +1,218 @@
+    createdAt:
+      match[1] != null && match[2] != null
+        ? new Date(`${match[1]}T${match[2].replace(/-/g, ":")}Z`)
+        : undefined,
+  };
+}
+
+export class Web3KeyStore implements ImportableKeyStore<KeyId, RawPrivateKey> {
+  readonly #passphraseEntry: PassphraseEntry;
+
+  readonly path: string;
+
+  constructor(options: Web3KeyStoreOptions) {
+    this.path = options.path ?? getDefaultWeb3KeyStorePath();
+    this.#passphraseEntry = options.passphraseEntry;
+  }
+
+  async *#listKeyFiles(): AsyncIterable<Dirent> {
+    let dir;
+    try {
+      dir = await fs.opendir(this.path);
+    } catch (e) {
+      if (
+        typeof e === "object" &&
+        e != null &&
+        "code" in e &&
+        e.code === "ENOENT"
+      ) {
+        return;
+      }
+      throw e;
+    }
+    for await (const dirEntry of dir) {
+      if (!dirEntry.isFile()) continue;
+      yield dirEntry;
+    }
+  }
+
+  async #getKeyPath(
+    keyId: KeyId,
+  ): Promise<{ path: string; keyId: string; createdAt?: Date } | undefined> {
+    for await (const dirEntry of this.#listKeyFiles()) {
+      const parsed = parseKeyFilename(dirEntry.name);
+      if (parsed != null && parsed.keyId === keyId) {
+        return { ...parsed, path: path.join(this.path, dirEntry.name) };
+      }
+    }
+    return undefined;
+  }
+
+  async *list(): AsyncIterable<AccountMetadata<KeyId>> {
+    for await (const dirEntry of this.#listKeyFiles()) {
+      const parsed = parseKeyFilename(dirEntry.name);
+      if (parsed == null) continue;
+      yield { ...parsed, metadata: undefined };
+    }
+  }
+
+  async get(
+    keyId: Readonly<KeyId>,
+  ): Promise<AccountRetrieval<KeyId, RawPrivateKey>> {
+    const keyPath = await this.#getKeyPath(keyId);
+    if (keyPath == null) return { result: "keyNotFound", keyId };
+    let json;
+    try {
+      json = await fs.readFile(keyPath.path, { encoding: "utf8" });
+    } catch (e) {
+      if (
+        e != null &&
+        typeof e === "object" &&
+        "code" in e &&
+        e.code === "ENOENT"
+      ) {
+        return { result: "keyNotFound", keyId };
+      }
+      return { result: "error", keyId, message: `${e}` };
+    }
+    if (!isKeystoreJson(json)) {
+      return { result: "error", keyId, message: "Invalid key file" };
+    }
+    let firstAttempt = true;
+    let account: KeystoreAccount;
+    while (true) {
+      const passphrase = await this.#passphraseEntry.authenticate(
+        keyId,
+        firstAttempt,
+      );
+      let currentProgress = 0;
+      try {
+        account = await decryptKeystoreJson(json, passphrase, (progress) => {
+          currentProgress = progress;
+        });
+      } catch (e) {
+        if (currentProgress <= 0) {
+          return { result: "error", keyId, message: `${e}` };
+        }
+        firstAttempt = false;
+        continue;
+      }
+      break;
+    }
+    return {
+      result: "success",
+      // TODO: Declare Web3Account
```

Comment:
Can you elaborate on this one? what kind of methods are on the table? I'm just curious.

Reply:
For the record, we discussed the possibility of carrying only the path of the keystore for the implementation, and different implementations of `PassphaseEntry` might be used for policy on what to do when key is intended to be used (so that passphrase can be off memory / can be requested once in a few minutes / or have a confirmation window, like the unix `sudo` command (or rather the Github sudo mode.))